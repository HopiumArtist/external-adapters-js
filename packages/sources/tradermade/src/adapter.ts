import { ExecuteWithConfig, Config, ExecuteFactory, AdapterRequest, AxiosResponse, MakeWSHandler } from '@chainlink/types'
import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { makeConfig, NAME, DEFAULT_WS_API_ENDPOINT } from './config'
import * as endpoints from './endpoint'


const customParams = {
  base: ['base', 'from', 'symbol', 'market'],
  to: false,
}

const handleBatchedRequest = (
  jobRunID: string,
  request: AdapterRequest,
  response: AxiosResponse<any>,
  resultPath: string,
) => {
  const payload: [AdapterRequest, number][] = []
  for (const pair of response.data.quotes) {
    const symbol = pair.base_currency
    const to = pair.quote_currency
    payload.push([
      {
        ...request,
        data: { ...request.data, from: symbol.toUpperCase(), to: to.toUpperCase() },
      },
      Requester.validateResultNumber(pair, [resultPath]),
    ])

  }
  return Requester.success(jobRunID, Requester.withResult(response, undefined, payload), true, ['from', 'to'])
}


export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
  const validator = new Validator(request, customParams)
  if (validator.error) throw validator.error

  Requester.logConfig(config)

  const jobRunID = validator.validated.id
  const symbol = validator.overrideSymbol(NAME)
  const to = (validator.validated.data.to || '')
  const pairArray = []
  console.log(to)
  for (const fromCurrency of formatArray(symbol)) {
    for (const toCurrency of formatArray(to)) {
      pairArray.push(`${fromCurrency.toUpperCase() + toCurrency.toUpperCase()}`)
    }
  }
  const currency = pairArray.toString()
  const params = {
    ...config.api.params,
    currency,
  }

  const options = { ...config.api, params }

  const response = await Requester.request(options)
  if (Array.isArray(symbol) || Array.isArray(to)) return handleBatchedRequest(jobRunID, request, response, 'mid')

  response.data.result = Requester.validateResultNumber(response.data, ['quotes', 0, 'mid'])
  return Requester.success(jobRunID, response, config.api.verbose, [{name: 'from',},{name: 'to',},])
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: Config): MakeWSHandler | undefined => {
  const getSubscription = (pair?: string) => {
    const defaultConfig = config || makeConfig()
    if (!pair) return
    const sub = {
      userKey: defaultConfig.wsApiKey,
      symbol: pair,
    }
    return sub
  }
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(input, endpoints.forex.customParams, {}, false)
    if (validator.error) return
    const base = validator.validated.data.base.toUpperCase()
    const quote = validator.validated.data.quote.toUpperCase()
    return `${base}${quote}`
  }
  return () => {
    const defaultConfig = config || makeConfig()
    return {
      connection: {
        url: defaultConfig.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
      },
      shouldNotServeInputUsingWS: (input: AdapterRequest) =>
        endpoints.forex.supportedEndpoints.indexOf(input.data.endpoint) === -1,
      subscribe: (input: AdapterRequest) => getSubscription(getPair(input)),
      unsubscribe: () => null, // Tradermade does not support unsubscribing.
      subsFromMessage: (message) => {
        if (!message.symbol) return undefined
        return getSubscription(message.symbol)
      },
      isError: () => false, // No error
      filter: (message: any) => !!message.mid,
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['mid'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
// format input as an array regardless of if it is a string or an array already
function formatArray(input: string | string[]) {
  let result = []
  if(typeof(input)==='string')
      result.push(input)
  else
      result = input
  return result
}
