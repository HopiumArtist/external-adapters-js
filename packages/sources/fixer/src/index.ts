import { expose } from '@chainlink/ea-bootstrap'
import { endpointSelector, makeExecute } from './adapter'
import { makeConfig, NAME } from './config'

const { server } = expose(NAME, makeExecute())
export { NAME, makeExecute, makeConfig, server }
