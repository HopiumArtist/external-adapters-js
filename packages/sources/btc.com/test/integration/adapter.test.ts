import { AdapterRequest } from '@chainlink/types'
import request from 'supertest'
import * as process from 'process'
import { server as startServer } from '../../src'
import * as nock from 'nock'
import * as http from 'http'
import { mockBalanceResponse, mockBlockResponse } from './fixtures'

describe('execute', () => {
  let server: http.Server
  const req = request('localhost:8080')
  beforeAll(async () => {
    process.env.CACHE_ENABLED = 'false'
    if (process.env.RECORD) {
      nock.recorder.rec()
    }
    server = await startServer()
  })
  afterAll((done) => {
    if (process.env.RECORD) {
      nock.recorder.play()
    }

    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
    server.close(done)
  })

  describe('balance endpoint', () => {
    const balanceRequest: AdapterRequest = {
      id: '1',
      data: {
        addresses: [
          {
            address: '3EyjZ6CtEZEKyc719NZMyWaJpJG5jsVJL1',
            coin: 'btc',
          },
        ],
        dataPath: 'addresses',
      },
    }

    it('should return success', async () => {
      mockBalanceResponse()

      const response = await req
        .post('/')
        .send(balanceRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('block endpoint', () => {
    const blockRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'difficulty',
      },
    }

    it('should return success', async () => {
      mockBlockResponse()

      const response = await req
        .post('/')
        .send(blockRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
