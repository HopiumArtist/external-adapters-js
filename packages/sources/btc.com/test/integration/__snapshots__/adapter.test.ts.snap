// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute block endpoint should return success 1`] = `
Object {
  "data": Object {
    "result": 22674148233453,
  },
  "jobRunID": "1",
  "result": 22674148233453,
  "statusCode": 200,
}
`;
